From cebbd5f26a68be33a71f89adf4fcc5ed02d0444b Mon Sep 17 00:00:00 2001
From: Yu Kuai <yukuai3@huawei.com>
Date: Fri, 2 Jun 2023 14:32:49 +0800
Subject: [PATCH 397/780] [Huawei] blk-io-hierarshy: add apis to alloc/free
 blk_io_hierarchy

Offering: HULK
hulk inclusion
category: bugfix
bugzilla: 188826

--------------------------------

blk_io_hierarchy is allocated/freed while request_queue is
allocated/freed.

Signed-off-by: Yu Kuai <yukuai3@huawei.com>
---
 block/blk-io-hierarchy/stats.c | 66 ++++++++++++++++++++++++++++++++++
 block/blk-io-hierarchy/stats.h | 13 +++++++
 block/blk-mq.c                 | 10 +++++-
 3 files changed, 88 insertions(+), 1 deletion(-)
 create mode 100644 block/blk-io-hierarchy/stats.c

diff --git a/block/blk-io-hierarchy/stats.c b/block/blk-io-hierarchy/stats.c
new file mode 100644
index 000000000000..6c87fd222282
--- /dev/null
+++ b/block/blk-io-hierarchy/stats.c
@@ -0,0 +1,66 @@
+// SPDX-License-Identifier: GPL-2.0
+/*
+ * Copyright (C) 2023. Huawei Technologies Co., Ltd. All rights reserved.
+ * This program is free software; you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License version 2 and
+ * only version 2 as published by the Free Software Foundation.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+ * GNU General Public License for more details.
+ */
+
+#include <linux/debugfs.h>
+
+#include "stats.h"
+
+static struct blk_io_hierarchy_stats *hierarchy_stats_alloc(void)
+{
+	struct blk_io_hierarchy_stats *stats =
+		kzalloc(sizeof(struct blk_io_hierarchy_stats), GFP_KERNEL);
+	if (!stats)
+		return ERR_PTR(-ENOMEM);
+
+	stats->stats = alloc_percpu(struct hierarchy_stats);
+	if (!stats->stats) {
+		kfree(stats);
+		return ERR_PTR(-ENOMEM);
+	}
+
+	return stats;
+}
+
+static void hierarchy_stats_free(struct blk_io_hierarchy_stats *stats)
+{
+	free_percpu(stats->stats);
+	kfree(stats);
+}
+
+int blk_io_hierarchy_stats_alloc(struct request_queue *q)
+{
+	struct blk_io_hierarchy_stats *stats;
+
+	if (!queue_is_mq(q))
+		return 0;
+
+	stats = hierarchy_stats_alloc();
+	if (IS_ERR(stats))
+		return PTR_ERR(stats);
+
+	stats->q = q;
+	q->io_hierarchy_stats = stats;
+
+	return 0;
+}
+
+void blk_io_hierarchy_stats_free(struct request_queue *q)
+{
+	struct blk_io_hierarchy_stats *stats = q->io_hierarchy_stats;
+
+	if (!stats)
+		return;
+
+	q->io_hierarchy_stats = NULL;
+	hierarchy_stats_free(stats);
+}
diff --git a/block/blk-io-hierarchy/stats.h b/block/blk-io-hierarchy/stats.h
index 87ee80152d22..2b022b23bae8 100644
--- a/block/blk-io-hierarchy/stats.h
+++ b/block/blk-io-hierarchy/stats.h
@@ -30,5 +30,18 @@ struct blk_io_hierarchy_stats {
 	struct hierarchy_stats __percpu *stats;
 };
 
+int blk_io_hierarchy_stats_alloc(struct request_queue *q);
+void blk_io_hierarchy_stats_free(struct request_queue *q);
+
+#else /* CONFIG_BLK_IO_HIERARCHY_STATS */
+
+static inline int blk_io_hierarchy_stats_alloc(struct request_queue *q)
+{
+	return 0;
+}
+
+static inline void blk_io_hierarchy_stats_free(struct request_queue *q)
+{
+}
 #endif /* CONFIG_BLK_IO_HIERARCHY_STATS */
 #endif /* BLK_IO_HIERARCHY_STATS_H */
diff --git a/block/blk-mq.c b/block/blk-mq.c
index 422c264e3040..dbe5c9739f57 100644
--- a/block/blk-mq.c
+++ b/block/blk-mq.c
@@ -40,6 +40,7 @@
 #include "blk-stat.h"
 #include "blk-mq-sched.h"
 #include "blk-rq-qos.h"
+#include "blk-io-hierarchy/stats.h"
 
 bool mq_unfair_dtag = true;
 module_param_named(unfair_dtag, mq_unfair_dtag, bool, 0444);
@@ -3271,6 +3272,8 @@ void blk_mq_release(struct request_queue *q)
 	struct blk_mq_hw_ctx *hctx, *next;
 	int i;
 
+	blk_io_hierarchy_stats_free(q);
+
 	queue_for_each_hw_ctx(q, hctx, i)
 		WARN_ON_ONCE(hctx && list_empty(&hctx->hctx_list));
 
@@ -3475,11 +3478,14 @@ struct request_queue *blk_mq_init_allocated_queue(struct blk_mq_tag_set *set,
 	/* mark the queue as mq asap */
 	q->mq_ops = set->ops;
 
+	if (blk_io_hierarchy_stats_alloc(q))
+		goto err_exit;
+
 	q->poll_cb = blk_stat_alloc_callback(blk_mq_poll_stats_fn,
 					     blk_mq_poll_stats_bkt,
 					     BLK_MQ_POLL_STATS_BKTS, q);
 	if (!q->poll_cb)
-		goto err_exit;
+		goto err_hierarchy_stats;
 
 	if (blk_mq_alloc_ctxs(q))
 		goto err_poll;
@@ -3538,6 +3544,8 @@ struct request_queue *blk_mq_init_allocated_queue(struct blk_mq_tag_set *set,
 err_poll:
 	blk_stat_free_callback(q->poll_cb);
 	q->poll_cb = NULL;
+err_hierarchy_stats:
+	blk_io_hierarchy_stats_free(q);
 err_exit:
 	q->mq_ops = NULL;
 	return ERR_PTR(-ENOMEM);
-- 
2.17.1

